### ISSUE TITLE
Keeping original dtypes when a file is partitioned by some columns

### ISSUE THREAD

**Issue by andrethrill** [dask/fastparquet#346]
**tldr**: if a save a parquet partitioned by some column, when I read back that same parquet, the dtype of the column used for partitioning is not kept: 

Original df dtypes:
```
ints                  int64
floats              float64
bools                 int64
fake_categorical      int64
dtype: object
```

Df dtypes read from disk:
```
ints                   int64
floats               float64
bools                  int64
fake_categorical    category
dtype: object
```

Is this the expected behaviour? Can the original dtypes be kept?

The code to reproduce this follows below:

# Imports

```python
import dask.dataframe as dd
import fastparquet as fp
import pandas as pd
import numpy as np
import os
import shutil
```

# Aux function

```python
def generate_df(size):
    d = {
        'dates' : pd.date_range('1980-01-01', periods=size, freq='1T'),
        'ints' : np.random.randint(0, size, size=size),
        'floats' : np.random.randn(size),
        'bools' : np.random.choice([0, 1], size=size),
        'fake_categorical' : np.random.choice([10, 20, 30, 40, 50], size=size) ,

    }

    df = pd.DataFrame(d)
    df.index = df.dates
    df.index.name = 'dates'
    df = df.drop('dates', axis=1)

    return df
```

# generate DF

```python
df = generate_df(1000)
```

```python
df.head()
```

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ints</th>
      <th>floats</th>
      <th>bools</th>
      <th>fake_categorical</th>
    </tr>
    <tr>
      <th>dates</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1980-01-01 00:00:00</th>
      <td>722</td>
      <td>2.081084</td>
      <td>1</td>
      <td>20</td>
    </tr>
    <tr>
      <th>1980-01-01 00:01:00</th>
      <td>526</td>
      <td>0.330207</td>
      <td>1</td>
      <td>50</td>
    </tr>
    <tr>
      <th>1980-01-01 00:02:00</th>
      <td>627</td>
      <td>-0.875390</td>
      <td>0</td>
      <td>30</td>
    </tr>
    <tr>
      <th>1980-01-01 00:03:00</th>
      <td>217</td>
      <td>0.997309</td>
      <td>1</td>
      <td>20</td>
    </tr>
    <tr>
      <th>1980-01-01 00:04:00</th>
      <td>258</td>
      <td>0.293013</td>
      <td>0</td>
      <td>40</td>
    </tr>
  </tbody>
</table>
</div>

```python
df.shape
```

    (1000, 4)

# Original df dtypes:

```python
df.dtypes
```

    ints                  int64
    floats              float64
    bools                 int64
    fake_categorical      int64
    dtype: object

# save to parquet

```python
fp.write('original_df.parquet', df, compression='snappy', file_scheme='hive', partition_on=['fake_categorical'])
```

# Create a copy

```python
!mkdir copied_df.parquet/
!cp -r original_df.parquet/* copied_df.parquet/
```

# Load copied parquet

```python
fpFile = fp.ParquetFile('copied_df.parquet/')
```

```python
df_new = fpFile.to_pandas(filters=[('fake_categorical','==',30)])
```

```python
df_new.head()
```

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ints</th>
      <th>floats</th>
      <th>bools</th>
      <th>fake_categorical</th>
    </tr>
    <tr>
      <th>dates</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>1980-01-01 00:02:00</th>
      <td>627</td>
      <td>-0.875390</td>
      <td>0</td>
      <td>30</td>
    </tr>
    <tr>
      <th>1980-01-01 00:07:00</th>
      <td>493</td>
      <td>1.213180</td>
      <td>1</td>
      <td>30</td>
    </tr>
    <tr>
      <th>1980-01-01 00:08:00</th>
      <td>90</td>
      <td>1.400426</td>
      <td>1</td>
      <td>30</td>
    </tr>
    <tr>
      <th>1980-01-01 00:09:00</th>
      <td>850</td>
      <td>-1.076183</td>
      <td>1</td>
      <td>30</td>
    </tr>
    <tr>
      <th>1980-01-01 00:11:00</th>
      <td>399</td>
      <td>-0.979044</td>
      <td>0</td>
      <td>30</td>
    </tr>
  </tbody>
</table>
</div>

```python
df_new.shape
```

    (177, 4)

# Df dtypes read from disk:

```python
df_new.dtypes
```

    ints                   int64
    floats               float64
    bools                  int64
    fake_categorical    category
    dtype: object

**Comment by martindurant** [comment]
This is expected behaviour: anything that is partitioning into a (small) number of directories must be catagory-like, as opposed to int64, which can have a large number of values. You can use the column as if it were numbers, though - do you have a specific use that isn't working because of this? Is the extra line to convert back to ints a pain?

**Comment by andrethrill** [comment]
No, it's not a pain. I just wanted to make sure this was expected. As from my perspective would make sense to keep dtypes equal to the way we originally saved it, as the user may not expect it. (but I completely understand having it as categorical is much more efficient and doesn't make much sense to keep a column with so few uniques as `int`)



As a continuation from above, I noticed something strange as well. If I read the original Dataframe as so:

```
fpFile = fp.ParquetFile('copied_df.parquet/')
df_new = fpFile.to_pandas(filters=[('fake_categorical','==',30)])
```

And then try to save it again:
```
fp.write('copied_df2.parquet/', df_new, compression='snappy',
         file_scheme='hive', partition_on=['fake_categorical'])
```
I get the error:

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-46-4ddafc76b8a6> in <module>()
      1 fp.write('copied_df2.parquet/', df_new, compression='snappy',
----> 2          file_scheme='hive', partition_on=['fake_categorical'])

~******/lib/python3.6/site-packages/fastparquet/writer.py in write(filename, data, row_group_offsets, compression, file_scheme, open_with, mkdirs, has_nulls, write_index, partition_on, fixed_text, append, object_encoding, times)
    870                     data[start:end], partition_on, filename, part, fmd,
    871                     compression, open_with, mkdirs,
--> 872                     with_field=file_scheme == 'hive'
    873                 )
    874                 fmd.row_groups.extend(rgs)

~******/lib/python3.6/site-packages/fastparquet/writer.py in partition_on_columns(data, columns, root_path, partname, fmd, compression, open_with, mkdirs, with_field)
    920     rgs = []
    921     for key in sorted(gb.indices):
--> 922         df = gb.get_group(key)[remaining]
    923         if not isinstance(key, tuple):
    924             key = (key,)

~******/lib/python3.6/site-packages/pandas/core/groupby/groupby.py in get_group(self, name, obj)
    875         inds = self._get_index(name)
    876         if not len(inds):
--> 877             raise KeyError(name)
    878 
    879         return obj._take(inds, axis=self.axis)

KeyError: 10
```

I guess this is because the column `fake_categorical` is like so:
```
$ df_new.fake_categorical.dtypes
CategoricalDtype(categories=[40, 10, 50, 20, 30], ordered=False)
```

But in `df_new` it only has instances with df.fake_categorical==30 because I've filtered it in `.to_pandas()` method.


Am I doing something wrong?

**Comment by martindurant** [comment]
You don't appear to be doing something wrong, but it is odd that one of the keys returned by `gb.indices` can't be referenced by `gb.get_group`. Maybe my understanding of what `indices` should be is wrong, maybe pandas changed, or maybe this is a bug upstream. It would be reasonable for fastparquet to catch this exception and skip the group. Also a check that the returned sub-dataframe contains rows might be made here, although I can't see why you would have empty ones.

**Comment by andrethrill** [comment]
I've done a quick test and if instead of doing (as it is currently being done in fastparquet):

```python
for key in sorted(gb.indices):
        df = gb.get_group(key)[remaining]
```

we do:

```
for key, _df in gb:
        df = _df[remaining]
```

It works.

What do you think about changing for something like this? (I can do a PR for this)

**Comment by martindurant** [comment]
sounds very reasonable
